### PR TITLE
Issue #209: Adds session feedback page

### PR DESCRIPTION
**Pull request by antonjb**
Would like feedback on the approach I have taken for this. Things that I
have introduced:

useSession hook
handles front end loading of Sessions. Added to Agenda and Feedback page
Mostly similar to what we have no but in hook form with reducer. Also
houses the loading code for SSR. I feel like this could be joined together
somehow because they're repeating. Any input would be good. Suspect CSR
supports async so it might basically be there

useSessionGroup hook
This is probably the biggest thing I need eyes on. The purpose is to keep
the sessions in groups and then be able to know what the past sessions,
previous sessions, current sessions, next sessions and future sessions are.
The gaps made this more difficult than I would have liked

How I see this being used:
* Show the previous sessions on the feedback page (done)
* Past sessions in a show/hide under the previous session on feedback page (done)
* Show the next sessions at the top of the agenda page without needing to hunt for it (todo)

The code for calculating it seems like it could use improvement. I have
tried several variations on this and trying to memoize it because of how
frequently it runs. The current version is great for memoization but maybe
a little gross. Setting up groups also leaves a little to be desired.

useForm hook
Hook that handles changing in form and calling a submit callback.

storageKey helper
Small helper function when we are using local storage

Finally is the feedback page. I have added test buttons here to be able to
jump between times to see the sessions. I have gone with the traffic like
approach but need input on how to display that.

TL;DR
* Hooks all the things
* Can I improve useSessionGroups? Particularly the logic calculating the state
* Feedback page - what do we think?

### CODE REVIEW

File: components/Feedback/FeedbackTimeTesting.tsx [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import { Moment } from 'moment'
```
Comment (antonjb) [inline components/Feedback/FeedbackTimeTesting.tsx]: This adds buttons on the feedback page to enable jumping around session times. It wont (shouldn't) show otherwise.

File: components/utils/useSessionGroups.ts [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+  ])
+  const isConferenceDay = dateTimeProvider.now().Value.isBetween(conferenceDate, conferenceEndDate)
+
+  let pastSessionsIndex: number = -1
+  let previousSessionIndex: number = -1
+  let nextSessionIndex: number = -1
+  let currentSessionIndex: number = -1
+  let futureSessionIndex: number = -1
+
+  for (let i = 0; i < allSessionGroups.length; i++) {
```
Comment (antonjb) [inline components/utils/useSessionGroups.ts]: I tried several variations with this. Grabbing the next session, mapping, reducer and all had their benefits/drawbacks. I don't know if this is the best approach or not but it's mostly clean

File: pages/feedback.tsx [hunk 3]
```diff
@@ -1,78 +1,254 @@
...
+enum StorageKeys {
+  FEEDBACK_ID = 'ddd-feedback-id',
+  FEEDBACK_GIVER = 'ddd-feedback-name',
+}
+
+interface FeedbackMetadataProps extends WithPageMetadataProps {
+  ssrSessions?: Session[]
+}
+
+function getLocalStoredName(instance: string): string {
```
Comment (antonjb) [inline pages/feedback.tsx]: Save their name so they don't have to repeat input

File: pages/feedback.tsx [hunk 4]
```diff
@@ -1,78 +1,254 @@
...
+  })
+
+  React.useEffect(() => {
+    if (!feedbackId) {
+      const feedbackUUID = uuid()
+      try {
+        localStorage.setItem(storageKey<StorageKeys>(conference.Instance, StorageKeys.FEEDBACK_ID), feedbackUUID)
+        setFeedbackId(localStorage.getItem(storageKey<StorageKeys>(conference.Instance, StorageKeys.FEEDBACK_ID)))
+      } catch (err) {
+        setFeedbackId(feedbackUUID)
```
Comment (antonjb) [inline pages/feedback.tsx]: The uuid that we are going to pass through with the feedback

File: pages/feedback.tsx [hunk 5]
```diff
@@ -1,78 +1,254 @@
...
+              <StyledRatingInput
+                type="radio"
+                name="rating"
+                value="-1"
+                id="negative"
+                checked={values.rating === '-1'}
+                onChange={handleChange}
+                required
+              />
+              <label htmlFor="negative">😞</label>
```
Comment (antonjb) [inline pages/feedback.tsx]: @aidanmorgan I don't know 🤷‍♂ what's best here?

File: pages/feedback.tsx [hunk 6]
```diff
@@ -1,78 +1,254 @@
...
+          </StyledForm>
+        )}
+      </StyledContainer>
+    </Page>
+  )
+}
+
+Feedback.getInitialProps = async ({ res, req }) => {
+  const dates = getConferenceDates(Conference, dateTimeProvider.now())
+  // DEBUG - Hope I don't check this in but I can't get here otherwise...
```
Comment (antonjb) [inline pages/feedback.tsx]: I couldn't get to the page without this. It's ok for now but it shouldn't make it in the end

Comment (robdmoore) [reply pages/feedback.tsx]: `Doesn't the testing control allow you to do this?`

Comment (antonjb) [reply pages/feedback.tsx]: I think you have to use a next link for it to work. the feedback page doesn't appear in the main nav. With that said, should it be in the main navigation?

Comment (antonjb) [reply pages/feedback.tsx]: I just found the link 😔

File: components/utils/useSessionGroups.ts [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+  futureSessionGroups?: SessionGroup[]
+}
+
+function getSessionById(sessions: Session[], ids: SessionId[]) {
+  return sessions.filter(session => ids.includes(session.Id))
+}
+
+// Creates the session groups.
+// Important: Order is assumed to be correct.
+function createSessionGroups(conferenceDate: Moment, sessions: Session[]): SessionGroup[] {
```
Comment (robdmoore) [inline components/utils/useSessionGroups.ts]: This stuff is specific to this year's conference so shouldn't appear in `components`. Suggestion would be to add it to the `Conference` object?

We'll then need to figure out how to incorporate that into our previous agenda stuff later in the year, but that should be fine.

File: components/utils/useSessionGroups.ts (hunk 2 above)
Comment (robdmoore) [reply components/utils/useSessionGroups.ts]: Yeah it's not the prettiest, but if it works and works reliably then I'm happy

File: pages/feedback.tsx (hunk 5 above)
Comment (robdmoore) [reply pages/feedback.tsx]: I think they look good

File: pages/feedback.tsx (hunk 6 above)
Comment (robdmoore) [reply pages/feedback.tsx]: As it's a separate page yes it should. We should also look at all the places we already had a link to feedback and change them, including from the agenda.

File: components/utils/useSessionGroups.ts (hunk 7 above)
Comment (antonjb) [reply components/utils/useSessionGroups.ts]: Good point, not the best home for this. Ideally it would set the agenda page up too but I'm not sure how to structure that yet with breaks etc. Perhaps more thought is needed on the data structure

**Comment by robdmoore** [comment]
For the purposes of getting this working let’s just pull it into conference object and leave agenda as is?

--



On 28 Jul 2019, at 1:25 pm, Anton <notifications@github.com<mailto:notifications@github.com>> wrote:


@antonjb commented on this pull request.

________________________________

In components/utils/useSessionGroups.ts<https://github.com/dddwa/dddperth-website/pull/214#discussion_r307985883>:

> +interface SessionGroups {
+  allSessionGroups: SessionGroup[]
+  currentSessionGroup?: SessionGroup
+  previousSessionGroup?: SessionGroup
+  nextSessionGroup?: SessionGroup
+  pastSessionGroups?: SessionGroup[]
+  futureSessionGroups?: SessionGroup[]
+}
+
+function getSessionById(sessions: Session[], ids: SessionId[]) {
+  return sessions.filter(session => ids.includes(session.Id))
+}
+
+// Creates the session groups.
+// Important: Order is assumed to be correct.
+function createSessionGroups(conferenceDate: Moment, sessions: Session[]): SessionGroup[] {


Good point, not the best home for this. Ideally it would set the agenda page up too but I'm not sure how to structure that yet with breaks etc. Perhaps more thought is needed on the data structure

**Comment by antonjb** [comment]
> For the purposes of getting this working let’s just pull it into conference object and leave agenda as is?

Yep, that was my intention. More a thought for future versions that it could support the agenda.
